rocAL/rocAL/source/pipeline/master_graph.cpp

```diff
@@ -580,108 +582,216 @@ void MasterGraph::stop_processing()
+    }
+    else if(reader_type == MetaDataReaderType::TF_DETECTION_META_DATA_READER)
+    {
+        std::vector<size_t> dims = { MAX_OBJECTS };
+        auto default_labels_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::INT32); // Create default labels Info
+        default_labels_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_labels_info.data_size());
+
+        dims = { MAX_OBJECTS, BBOX_COUNT };
+        auto default_bbox_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::FP32); // Create default Bbox Info
+        default_bbox_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_bbox_info.data_size());
+
+        for(unsigned i = 0; i < _user_batch_size; i++)
+        {
+            auto labels_info = default_labels_info;
+            auto bbox_info = default_bbox_info;
+            _labels_tensor_list.push_back(new rocalTensor(labels_info));
+            _bbox_tensor_list.push_back(new rocalTensor(bbox_info));
+        }
+        _metadata_output_tensor_list.emplace_back(&_labels_tensor_list);
+        _metadata_output_tensor_list.emplace_back(&_bbox_tensor_list);
+    }
+
+    _ring_buffer.init_metadata(RocalMemType::HOST, _meta_data_buffer_size);
+
+    return _metadata_output_tensor_list;
 }
 
-MetaDataBatch * MasterGraph::create_label_reader(const char *source_path, MetaDataReaderType reader_type)
+std::vector<rocalTensorList *> MasterGraph::create_label_reader(const char *source_path, MetaDataReaderType reader_type)
 {
-    if( _meta_data_reader)
+    if(_meta_data_reader)
         THROW("A metadata reader has already been created")
-    MetaDataConfig config(MetaDataType::Label, reader_type, source_path);
-    _meta_data_reader = create_meta_data_reader(config);
-    _meta_data_reader->init(config);
-    _meta_data_reader->read_all(source_path);
     if (_augmented_meta_data)
         THROW("Metadata can only have a single output")
-    else
-        _augmented_meta_data = _meta_data_reader->get_output();
-    return _meta_data_reader->get_output();
+
+    MetaDataConfig config(MetaDataType::Label, reader_type, source_path);
+    _meta_data_reader = create_meta_data_reader(config, _augmented_meta_data);
+    _meta_data_reader->read_all(source_path);
+
+    std::vector<size_t> dims = {1};
+    auto default_labels_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::INT32); // Create default labels Info
+    default_labels_info.set_metadata();
+    _meta_data_buffer_size.emplace_back(_user_batch_size * sizeof(vx_int32));
+
+    for(unsigned i = 0; i < _user_batch_size; i++)
+    {
+        auto info = default_labels_info;
+        _labels_tensor_list.push_back(new rocalTensor(info));
+    }
+    _ring_buffer.init_metadata(RocalMemType::HOST, _meta_data_buffer_size);
+    _metadata_output_tensor_list.emplace_back(&_labels_tensor_list);
```

Comment:
why can't we return the labels_tensor_list directly from here?

Reply:
All the metadata readers will be returning a vector of RocalTensorList(RocalMetadata), where each tensor list corresponds to the metadata.
In DALI all the readers return the tensorList for each metadata, so in order to make our API similar to DALI we are returning the RocalMetaData.